### PR TITLE
fix: optimize `emplace_back()` using `reserve()`

### PR DESCRIPTION
**Pull request by GermanAizek**
@variar,
You may have forgotten to add reserve before using `emplace_back`, because it works almost the same as `push_back`.

### CODE REVIEW

**Comment by variar** [comment]
Hi, `reserve` indeed can improve things in several places. However, adding reserve in argument parsing code or settings loading increases code size without any significant performance gains. I'd prefer to have simplier code in parts not related to searching and rendering results.

**Comment by GermanAizek** [comment]
@variar, Specify where to undo commit changes.

File: src/app/cli.h [hunk 1]
```diff
@@ -152,6 +152,7 @@ struct CliParameters {
             }
         }
 
+        filenames.reserve( static_cast<size_t>( parser.positionalArguments().count() ) );
```
Comment (variar) [inline src/app/cli.h:155]: Usually there are not that many files on command line. Reserve here not needed.

File: src/settings/src/persistentinfo.cpp [hunk 2]
```diff
@@ -186,7 +186,8 @@ void PersistentInfo::UpdateSettings()
             const auto id = generateIdFromUuid();
             sessionSettings_->beginGroup( "OpenFiles" );
             std::vector<std::tuple<QString, uint64_t, QString>> openFiles;
-            int size = sessionSettings_->beginReadArray( "openFiles" );
+            const int size = sessionSettings_->beginReadArray( "openFiles" );
```
Comment (variar) [inline src/settings/src/persistentinfo.cpp:189]: This place is reading settings, so not performance critical. Can stay without reserve

File: src/ui/src/favoritefiles.cpp [hunk 3]
```diff
@@ -80,7 +80,8 @@ void FavoriteFiles::retrieveFromStorage( QSettings& settings )
     if ( settings.contains( "FavoriteFiles/version" ) ) {
         settings.beginGroup( "FavoriteFiles" );
         if ( settings.value( "version" ).toInt() == FavoriteFilesVersion ) {
-            int size = settings.beginReadArray( "favorites" );
+            const int size = settings.beginReadArray( "favorites" );
```
Comment (variar) [inline src/ui/src/favoritefiles.cpp:83]: Another reading settings place. Let's remove reserve

File: src/ui/src/mainwindow.cpp [hunk 4]
```diff
@@ -1947,6 +1947,7 @@ void MainWindow::writeSettings()
     std::vector<
         std::tuple<const ViewInterface*, uint64_t, std::shared_ptr<const ViewContextInterface>>>
         widget_list;
+    widget_list.reserve( static_cast<size_t>( mainTabWidget_.count() ) );
```
Comment (variar) [inline src/ui/src/mainwindow.cpp:1950]: Another settings place. Let's remove reserve

File: src/ui/src/session.cpp [hunk 5]
```diff
@@ -147,6 +147,7 @@ std::vector<WindowSession> Session::windowSessions()
     const auto& sessionWindows = session.windows();
 
     std::vector<WindowSession> windows;
+    windows.reserve( static_cast<size_t>( sessionWindows.size() ) );
```
Comment (variar) [inline src/ui/src/session.cpp:150]: Another settings place. Let's remove reserve

File: src/ui/src/session.cpp [hunk 6]
```diff
@@ -189,8 +190,11 @@ WindowSession::restore( const std::function<ViewInterface*()>& view_factory,
     const auto& session = SessionInfo::getSynced();
 
     std::vector<SessionInfo::OpenFile> session_files = session.openFiles( windowId_ );
-    LOG_DEBUG << "Session returned " << session_files.size();
+    const size_t count = session_files.size();
```
Comment (variar) [inline src/ui/src/session.cpp:193]: Another settings place. Let's remove reserve

File: src/ui/src/sessioninfo.cpp [hunk 7]
```diff
@@ -34,7 +34,8 @@ void SessionInfo::retrieveFromStorage( QSettings& settings )
 
     if ( settings.value( "version", 0 ).toInt() == SESSION_VERSION ) {
         windows_.clear();
-        const auto windowsCount = settings.beginReadArray( "windows" );
+        const int windowsCount = settings.beginReadArray( "windows" );
```
Comment (variar) [inline src/ui/src/sessioninfo.cpp:37]: Another settings place. Let's remove reserve

File: src/ui/src/sessioninfo.cpp [hunk 8]
```diff
@@ -44,7 +45,8 @@ void SessionInfo::retrieveFromStorage( QSettings& settings )
             if ( settings.contains( "OpenFiles/version" ) ) {
                 settings.beginGroup( "OpenFiles" );
                 if ( settings.value( "version" ).toInt() == OPENFILES_VERSION ) {
-                    int size = settings.beginReadArray( "openFiles" );
+                    const int size = settings.beginReadArray( "openFiles" );
```
Comment (variar) [inline src/ui/src/sessioninfo.cpp:48]: Another settings place. Let's remove reserve

File: src/ui/src/tabnamemapping.cpp [hunk 9]
```diff
@@ -84,7 +84,8 @@ void TabNameMapping::retrieveFromStorage( QSettings& settings )
     if ( settings.contains( "TabNameMapping/version" ) ) {
         settings.beginGroup( "TabNameMapping" );
         if ( settings.value( "version" ).toInt() == TABNAMEMAPPING_VERSION ) {
-            int size = settings.beginReadArray( "tabNames" );
+            const int size = settings.beginReadArray( "tabNames" );
```
Comment (variar) [inline src/ui/src/tabnamemapping.cpp:87]: Another settings place. Let's remove reserve

**Review by variar — CHANGES_REQUESTED** [review]
I've indicated places where in my opinion reserve is not adding enough value. Please, remove it.

**Comment by GermanAizek** [comment]
@variar,
I didn't check all `reserve()` before `push_back()`, after fix this PR I need to do new PR with `reserve()` with `push_back()`?
Do you mind pointing out again in other PR where `reserve()` should not be used because of the small count loop steps?

**Comment by GermanAizek** [comment]
I am an active user your program, and I would like to speed up what is possible.
On my server, the longest process is file indexing. It often freezes during a quick search, and application does not respond.

**Comment by GermanAizek** [comment]
On average, I look at logs in sizes from 50 GB to 150 GB.

**Comment by GermanAizek** [comment]
@variar,
Can you tell me if file indexing is multithreaded?

File: src/app/cli.h (hunk 1 above)
Comment (GermanAizek) [reply src/app/cli.h]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/settings/src/persistentinfo.cpp (hunk 2 above)
Comment (GermanAizek) [reply src/settings/src/persistentinfo.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/ui/src/favoritefiles.cpp (hunk 3 above)
Comment (GermanAizek) [reply src/ui/src/favoritefiles.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/ui/src/mainwindow.cpp (hunk 4 above)
Comment (GermanAizek) [reply src/ui/src/mainwindow.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/ui/src/session.cpp (hunk 5 above)
Comment (GermanAizek) [reply src/ui/src/session.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/ui/src/session.cpp (hunk 6 above)
Comment (GermanAizek) [reply src/ui/src/session.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/ui/src/sessioninfo.cpp (hunk 7 above)
Comment (GermanAizek) [reply src/ui/src/sessioninfo.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/ui/src/sessioninfo.cpp (hunk 8 above)
Comment (GermanAizek) [reply src/ui/src/sessioninfo.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

File: src/ui/src/tabnamemapping.cpp (hunk 9 above)
Comment (GermanAizek) [reply src/ui/src/tabnamemapping.cpp]: fix it https://github.com/variar/klogg/commit/a4f9cad9eceff87a667f1c6cec939f383153a6e3

**Comment by variar** [comment]
Thanks for your contributions ) My average use case was 1-5Gb logs. Speeding up on 50Gb size is a good challenge. I am looking forward to seeing results.

> after fix this PR I need to do new PR with reserve() with push_back()?

Can be in this PR.

> Do you mind pointing out again in other PR where `reserve()` should not be used because of the small count loop steps?

Sure. I can check that.

> On my server, the longest process is file indexing. It often freezes during a quick search, and application does not respond.

Indexing is done in separate threads, it should not freeze the UI. The main regex search is also done in separate threads. However, if regex search produces a lot of matches, then the UI could freeze updating the "matches overview" -- the area to the right of main text view that shows red and blue lines to indicate where in a file tha matches are located.  Disabling the overview could help. 

The quick find (the one that can be called via Ctrl+F shortcut) is also done in a separate thread, but I've seen it freezing sometimes.

Indexing is multithreaded. There is one thread that reads blocks of data from a file and puts them into a queue. Then there are multiple threads (one per core usually) that get blocks from the queue and parse it to get end of line offsets. 
And there is the thread that sequences the results from parsing threads. There is a setting "index file read buffer" that can change the size of single block.

Main regex search is also mutlithreaded. There is a thread that reads blocks of lines from the file and puts them into a queue. There are several matching threads (one per core usually) that get a block of lines from the queue and output matching line numbes. And finally there is a thread that combines results of matching threads.  The number of lines in a block can be changed in advanced settings tab.

This mutlithreading setup is mostly done using Intel Threadin Buiding Blocks flow graph library.

One more idea about the freeze -- check that klogg logging is disabled. Logging could be a point of syncronization for all threads, I've seen it significantly reducing search performance.

**Comment by GermanAizek** [comment]
> Indexing is multithreaded.


I understand correctly that everything rests on SSD limit and I need to raise RAID 1+0 (RAID 10) for double bandwidth?

**Comment by variar** [comment]
> > Indexing is multithreaded.
> 
> I understand correctly that everything rests on SSD limit and I need to raise RAID 1+0 (RAID 10) for double bandwidth?

The indexing part should indeed be IO bound, so increasing IO bandwidth could help with inital file loading. Searching on the other hand is more cpu intensive.

**Pull request merged**